packages/destination-actions/src/destinations/ortto/ortto-client.ts

```
@@ -0,0 +1,209 @@
+import { AudienceList, Audience } from './types'
+
+export const API_VERSION = 'v1'
+
+export const Errors: Record<string, string> = {
+  MissingIDs: `Either user ID or anonymous ID must be specified`,
+  InvalidAPIKey: `Invalid API key`,
+  MissingEventName: `Missing event name`,
+  MissingAudienceName: `Missing audience name`,
+  MissingAudienceId: `Missing audience Id`,
+  MissingContactID: `At least one contact Id must be provided`
+}
+export default class OrttoClient {
+  request: RequestClient
+  constructor(request: RequestClient) {
+    this.request = request
+  }
+
+  upsertContacts = async (settings: Settings, payloads: UpsertContactPayload[], hookAudienceID: string) => {
+    const cleaned = []
+    for (let i = 0; i < payloads.length; i++) {
+      const event = payloads[i]
+      if (!event.anonymous_id && !event.user_id) {
+        throw new PayloadValidationError(Errors.MissingIDs)
+      }
+      if (!event.audience_id) {
+        event.audience_id = hookAudienceID
+      }
+      cleaned.push(cleanObject(event))
+    }
+    if (cleaned.length == 0) {
+      return
+    }
+    const url = this.getEndpoint(settings.api_key).concat('/identify')
+    return this.request(url, {
+      method: 'POST',
+      json: cleaned
+    })
+  }
+
+  sendActivities = async (settings: Settings, payloads: EventPayload[], hookAudienceID: string) => {
+    const filtered = []
+    for (let i = 0; i < payloads.length; i++) {
+      const event = payloads[i]
+      if (!event.anonymous_id && !event.user_id) {
+        throw new PayloadValidationError(Errors.MissingIDs)
+      }
+      if (!event.event || event.event.trim() === '') {
+        throw new PayloadValidationError(Errors.MissingEventName)
+      }
+      if (event.namespace === 'ortto.com') {
+        continue
+      }
+      if (!event.audience_id) {
+        event.audience_id = hookAudienceID
+      }
+      filtered.push(cleanObject(event))
+    }
+    if (filtered.length == 0) {
+      return
```

Maybe throw an error if there are no valid payloads?

---

I don’t think we should. If an individual payload is truly invalid, we already return a `PayloadValidationError`. In this case, it’s reasonable to silently drop track events sent by Ortto— as users can send activities from Ortto to Segment, and silent dropping will prevent infinite flow of data. It's a valid scenario imo, not an error state.